pycqed/scripts/Spectroscopy.py

```diff
@@ -0,0 +1,281 @@
+import lmfit
+
+class Spectroscopy(ba.BaseDataAnalysis):
+    def __init__(self, t_start,
+                 options_dict,
+                 t_stop = None,
+                 extract_only=False,
+                 auto=True):
+        super(Spectroscopy, self).__init__(t_start, t_stop=t_stop,
+                                             options_dict=options_dict,
+                                             extract_only=extract_only)
+        self.extract_fitparams = self.options_dict.get('fitparams',True)
+        self.params_dict = {'freq':'sweep_points',
+                            'amp':'amp',
+                            'phase':'phase'}
+        if self.extract_fitparams:
+            self.params_dict.update({'fitparams':'fit_params'})
+
+        self.numeric_params = ['freq', 'amp', 'phase']
+        if 'qubit_label' in self.options_dict:
+            self.labels.extend(self.options_dict['qubit_label'])
+        if auto is True:
+            self.run_analysis()
+
+    def process_data(self):
+        self.freq_label = 'Frequency (GHz)'
+        self.amp_label = 'Transmission amplitude (arb. units)'
+
+        self.phase_label = 'Transmission phase (degrees)'
+        self.freq_range = self.options_dict.get('freq_range',None)
+        self.amp_range = self.options_dict.get('amp_range',None)
+        self.phase_range = self.options_dict.get('phase_range',None)
+        self.plotsize = self.options_dict.get('plotsize',(8,5))
+        if len(self.data_dict['timestamps']) == 1:
+            self.plot_frequency = np.squeeze(self.data_dict['freq'])
+            self.plot_amp = np.squeeze(self.data_dict['amp'])
+            self.plot_phase = np.squeeze(self.data_dict['phase'])
+            print(self.plot_frequency.shape, self.plot_amp.shape, self.plot_phase.shape)
+
+        else:
+            #TRANSPOSE ALSO NEEDS TO BE CODED FOR 2D
+            self.plot_xvals = self.options_dict.get('xvals',np.array([[tt] for tt in range(len(self.data_dict['timestamps']))]))
+            self.plot_xwidth = self.options_dict.get('xwidth',None)
+            self.plot_xlabel = self.options_dict.get('xlabel','Scan number')
+            if self.plot_xwidth == 'auto':
+                x_diff = np.diff(np.ravel(self.plot_xvals))
+                dx1 = np.concatenate(([x_diff[0]],x_diff))
+                dx2 = np.concatenate((x_diff,[x_diff[-1]]))
+                self.plot_xwidth = np.maximum(dx1,dx2)
+                self.plot_frequency = np.array([[tt] for tt in self.data_dict['freq']])
+                self.plot_phase = np.array([[tt] for tt in self.data_dict['phase']])
+                self.plot_amp = np.array([np.array([tt]).transpose() for tt in self.data_dict['amp']])
+
+            else:
+                self.plot_frequency = self.data_dict['freq']
+                self.plot_phase = self.data_dict['phase']
+                self.plot_amp = self.data_dict['amp']
+
+
+
+
+
+
+
+    def prepare_plots(self):
+
+        if len(self.data_dict['timestamps']) == 1:
```

Comment:
I guess this is how you deal with "single_timestamp" analyses. I'll see if I can adopt that in some better way. I ran into quite some trouble with the default plot_line functionality.